## Problem 4: Palindromic Numbers

A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.

In [1]:
import numpy as np

In [2]:
''' 
Determines if number is a palindrome. Does this by converting the number to a string, flipping the string and comparing
it to the original string. If each element matches its flipped counterpart, then the number is a palindrome.

eg:
>>> is_palindrome(1000)
>>> False
'''

def is_palindrome(number : 'int | str') -> bool:
    
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Determines if number is a palindrome. Does this by converting the number to a string, flipping the string and comparing
        it to the original string. If each element matches its flipped counterpart, then the number is a palindrome.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int | str : number : The number which is being tested.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : bool : Whether the number is a palindrome.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> is_palindrome(1000)
        >>> False

        >>> is_palindrome('123321')
        >>> True

        >>> is_palindrome(123321)
        >>> True
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Test for palindrome by turning number into a string and comparing it to it's reverse.
    # [::-1] = start at the end, end at the start, and move in steps of -1 = reverse
    return str(number) == str(number)[::-1]

In [3]:
''' 
Finds all the palindromic numbers that are products of 3-digit numbers, and records the numbers needed to generate the products.

eg:
>>> find_palindromes()
>>> ([10201, 11211, 12221, ...], [[101, 101], [101, 111], [101, 121], ...])
'''

def find_palindromes() -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Finds all the palindromic numbers that are products of 3-digit numbers, and records the numbers needed 
        to generate the products.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : Returns two lists :
            : list : palindromes : The list of all palindrome numbers.
            : list : multiples : The two numbers used in the product of the palindromes.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> find_palindromes()
        >>> ([10201, 11211, 12221, ...], [[101, 101], [101, 111], [101, 121], ...])
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    palindromes = []
    multiples = []
    
    # Search from 100 up to 999.
    for first_multiplier in range(100, 1000):
        for second_multiplier in range(100, 1000):
            
            # Calculate the product and test if it is a palindrome.
            product = first_multiplier * second_multiplier
            if is_palindrome(product):
                palindromes.append(product)
                multiples.append([first_multiplier, second_multiplier])
    
    return palindromes, multiples

In [4]:
results, multiples = find_palindromes()

solution = np.max(results)
solution_multiples = multiples[results.index(solution)]

In [5]:
print(f"The largest palindrome made from the product of two 3-digit numbers = {solution_multiples[0]} x {solution_multiples[1]} = {solution:,}")

The largest palindrome made from the product of two 3-digit numbers = 913 x 993 = 906,609


In [6]:
''' 
Finds all the palindromic numbers that are products of an x-digit number multiplied by a y-digit number, and records the 
numbers needed to generate the products.

eg:
>>> find_palindromes_generic(2, 4)
>>> ([11011, 12221, 13431, ...], [[11, 1001], [11, 1111], [11, 1221], ...])
'''

def find_palindromes_generic(first_number_digits, second_number_digits):
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Finds all the palindromic numbers that are products of an x-digit number multiplied by a y-digit number, 
        and records the numbers needed to generate the products.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : Returns two lists :
            : list : palindromes : The list of all palindrome numbers.
            : list : multiples : The two numbers used in the product of the palindromes.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> find_palindromes_generic(2, 4)
        >>> ([11011, 12221, 13431, ...], [[11, 1001], [11, 1111], [11, 1221], ...])
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(first_number_digits, int): raise ValueError('Please enter an integer > 0 for the first_number_digits argument.')
    if first_number_digits < 1: raise ValueError('Please enter an integer > 0 for the first_number_digits argument.')
    if not isinstance(second_number_digits, int): raise ValueError('Please enter an integer > 0 for the second_number_digits argument.')
    if second_number_digits < 1: raise ValueError('Please enter an integer > 0 for the second_number_digits argument.')
        
    palindromes = []
    multiples = []
    
    # Search from 1XXX to 9XXX.
    for first_multiplier in range(10**(first_number_digits-1), 10**(first_number_digits)):
        
        # Search from 1YYY to 9YYY.
        for second_multiplier in range(10**(second_number_digits-1), 10**(second_number_digits)):
            
            # Calculate the product and test if it is a palindrome.
            product = first_multiplier * second_multiplier
            if is_palindrome(product):
                palindromes.append(product)
                multiples.append([first_multiplier, second_multiplier])
                
    return palindromes, multiples

In [7]:
first_number_digits = 2
second_number_digits = 4
results, multiples = find_palindromes_generic(first_number_digits, second_number_digits)

solution = np.max(results)
solution_multiples = multiples[results.index(solution)]

print(f"The largest palindrome made from the product of a {first_number_digits}-digit number and a {second_number_digits}-digit number = {solution_multiples[0]} x {solution_multiples[1]} = {solution:,}")

The largest palindrome made from the product of a 2-digit number and a 4-digit number = 99 x 9911 = 981,189


### Problem 4 Solution: 

The largest palindrom made from the product of two 3-digit numbers = 913 x 993 = 906,609.